docs/tendermint-core/p2p/peer_manager.md

```diff
@@ -0,0 +1,439 @@
+
+## Peer life cycle
+
+For implementing the connection policy, the peer manager keeps track of the
+state of peers and manages their life-cycle.
+The life cycle of a peer is summarized in the picture below.
+The circles represent _states_ of a peer and the rectangles represent
+_transitions_.
+All transitions are performed by the `Router`, by invoking methods of the peer
+manager with corresponding names.
+Normal transitions are represented by green arrows, while red arrows represent
+alternative transitions taken in case of errors.
+
+<img src="pics/p2p-v0.35-peermanager.png" alt="peer life cycle" title="" width="600px" name="" align="center"/>
+
+### Candidate peer
+
+The initial state of a peer in the peer manager.
+
+A `Candidate` peer may become an actual peer, to which the node is connected.
+We do not use candidate to refer to a peer to which we are connected, nor to
+a peer we are attempting to connect.
+
+Candidate peers from which the router recently disconnected or failed to dial
+are, during a certain period, not eligible for establishing connections.
+This scenario is represented by the `Frozen Candidate` state.
+
+### DialNext transition
+
+This state transition is performed when the [connection policy](#connection-policy)
+determines that the node should try to establish a connection with a peer.
+
+The peer manager selects the [best-ranked](#peer-ranking) peer which is in the
+[`Candidate`](#candidate-peer) state and provides it to the router.
+As the router is supposed to dial the peer, the peer manager sets the peer to
+the [dialing](#dialing-peer) state.
+
+Dialing a candidate peer may have become possible because the peer manager
+has found a connection slot to [upgrade](#slots-upgrades) for given room to the
+selected candidate peer.
+If this is the case, the peer occupying this connection slot is set to the
+[upgrading state](#upgrading-peer), so that it will be evicted once the
+connection to the candidate peer is successfully established.
+
+### Dialing peer
+
+A peer that has been returned to the router as the next peer to dial.
+The router should be attempting to connect to this peer.
+
+A peer in `Dialing` state it is not considered as a candidate peer.
+
+### Dialed transition
+
+This transition is performed when the node establishes an outgoing connection
+with a peer.
+This means that the peer manager has provided this peer to the router as the
+[next peer to dial](#dialnext-transition), and the router has dialed and
+successfully established a connection with the peer.
+The peer is thus expected to be in the [`Dialing`](#dialing-peer) state.
+
+It may occur, however, that when this transition is invoked the peer is already
+in the [`Connected`](#connected-peer) state.
+The most likely reason is that the router, while dialing this peer, has also
+accepted an incoming connection from the same peer.
+In this case, the transition fails, indicating to the router that is should
+close the newly established connection.
```

Comment:
This is not the first "issue" we noticed by reading this spec. We should add a section, or even a document, listing the issues we found.

What you described actually happens, according to @tychoish. This is not the "fault" of the peer manager, as it is thought to avoid this scenario using locks, but by running the same algorithm independently, it happens that two nodes do not connect to each other because they are running the same algorithm at the same time.

Reply:
Exactly. We should have a mechanism to deal with this.

Reply:
We should build this scenarios properly. Of course, everything depends on the timing of transitions at each side, but we should be able to identify the critical "paths" there.

Reply:
@jmalicevic, I didn't really get the scenario. A node receives the report that a peer was disconnected, and this node tries to connect again to the peer, correct? Who receives the following error, and what is the specific error logged?

Reply:
So what I see in the logs of Node Y is :
- Error connecting to node X, node disconnected.
- Refused connection from node X, node already connected.

Node Y halts. I will dig out the exact error as soon as I get it again.

Reply:
Wow, we should discuss this synchronously, but it looks like a bug, not a design flaw.

Reply:
Yup, I don't think it is a design flaw in principle but probably the result of a logical race between marking a peer as disconnected and connecting. It is also not obvious where this happens as a from what I saw,these transitions are surrounded by a mutex.